ipa-core/src/ff/mod.rs

```diff
@@ -93,6 +93,29 @@ pub trait ArrayAccess {
     fn iter(&self) -> Self::Iter<'_>;
 }
 
+// There are currently a few places that the by-reference access of `ArrayAccessReplicated` is
+// required, and many places that the by-value access of `ArrayAccess` is used.
+//
+// If all uses of `ArrayAccess` were changed to operate by-reference, the replicated vs. not
+// distinction may not be a good reason to have separate traits. Uses of the replicated API would
+// just specify an additional `BorrowReplicated` bound. Note that there is at least one
+// `ArrayAccess` implementation where a straightforward adaptation will not make it work with
+// references: Galois fields, which have `ArrayAccess::Output = bool`. It probably would be possible
+// to provide by-reference access for Galois fields using the bit-reference functionality in the
+// bitvec crate.
+//
+// No reason this shouldn't have iter too, but deferring that because it's not needed yet, and will
+// be messy to support for Galois fields because of the `bool` value type.
+pub trait ArrayAccessReplicated {
+    type Ref<'a>: BorrowReplicated
+    where
+        Self: 'a;
```

Comment:
for my own curiosity, why is it better than `BorrowReplicated<'a>` ?

Reply:
I think the `BorrowReplicated` trait could have a lifetime parameter, but it's not necessary -- the implicit lifetime parameter in the reference argument to the trait methods is sufficient. Even if the trait were `BorrowReplicated<'a>`, the lifetime parameter would still need to propagate to something here (either the `Ref<'a>` GAT or a lifetime parameter on `ArrayAccessReplicated`. Just saying `BorrowReplicated<'a>` isn't meaningful and wouldn't work, because the `'a` lifetime would be undefined.

The `where Self: 'a` is something the compiler insists on -- it ensures that an impl like the one for `BitDecomposed<S>` with `Ref<'a>` of `BitDecomposedIndex<'a, S>(&'a S)` is possible. Using the type `&'a S` creates an implied bound `S: 'a` for that implementation. Absent `where S: 'a` in the trait definition, the implementation would have stricter bounds than the definition. This is discussed in more detail in rust issue 87479.